File: game/IdleMarket/Assets/Scripts/Equipments/EquipmentGenerator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

public static class EquipmentGenerator
{
    public static Equipment GenerateRandomEquipment(int rarity)
    {
        EquipmentType randomType = (EquipmentType)UnityEngine.Random.Range(
            0, Enum.GetValues(typeof(EquipmentType)).Length);

        return GenerateEquipment(rarity, randomType);
    }

    public static Equipment GenerateEquipment(int rarity, EquipmentType type)
    {
        StatType mainStat = EquipmentHelper.GetMainStat(type);

        List<SubStat> subStats = GetSubStatsByRarity(rarity);

        float mainStatValue = EquipmentHelper.GetMainStatRange(rarity).Roll();
        int rating = GenerateRating(mainStatValue, subStats, rarity);

        return new Equipment(
            type, rarity, mainStat, mainStatValue, rating, subStats
        );   
    }

    private static List<SubStat> GetSubStatsByRarity(int rarity)
    {
        List<SubStat> subStats = new();

        List<StatType> pool = new()
        {
            StatType.Health, StatType.Attack, StatType.Defense,
            StatType.CriticalChance, StatType.CriticalDamage
        };

        int subStatCount = rarity - 1;

        for (int i = 0; i < subStatCount; i++)
        {
            int index = UnityEngine.Random.Range(0, pool.Count);
            StatType subStatType = pool[index];
            pool.RemoveAt(index);
            
            StatRange statRange = EquipmentHelper.GetSubStatRange(subStatType);

            float subStatValue = statRange.Roll();
            subStats.Add(new SubStat(subStatType, subStatValue));
        }

        return subStats;
    }

    private static int GenerateRating(float mainStatValue, List<SubStat> subStats, int rarity)
    {
        List<float> scores = new();

        StatRange mainStatRange = EquipmentHelper.GetMainStatRange(rarity);

        scores.Add(mainStatRange.Normalize(mainStatValue) * 100f);

        foreach (SubStat subStat in subStats)
        {
            StatRange subStatRange = EquipmentHelper.GetSubStatRange(subStat.statType);
            scores.Add(subStatRange.Normalize(subStat.statValue) * 100f);
        }

        return Mathf.RoundToInt(scores.Average());
    }
}

File: backend/src/game/equipmentGenerator.ts
import { getMainStat, getMainStatRange, getSubStatRange } from "./equipmentHelper.js"
import { StatType, EquipmentType } from "../../generated/prisma/enums.js"
import { rangeInt } from "./randomHelper.js"

type Equipment = {
    type: EquipmentType, 
    rarity: number, 
    mainStat: StatType, 
    mainStatValue: number, 
    rating: number, 
    subStats: SubStat[]
}

type SubStat = { statType: StatType, statValue: number}

export function generateEquipment(rarity: number): Equipment {
    const type = getRandomEquipmentType()
    
    const mainStat = getMainStat(type)
    const mainStatValue = getMainStatRange(rarity).roll()

    const subStats = getSubStats(rarity)
    const rating = generateRating(mainStatValue, subStats, rarity)

    return {
        type: type,
        rarity: rarity,
        mainStat: mainStat,
        mainStatValue: mainStatValue,
        rating: rating,
        subStats: subStats
    }
}

function getSubStats(rarity: number): SubStat[] {
    const subStats: SubStat[] = []

    const pool = [
        StatType.Health, StatType.Attack, StatType.Defense,
        StatType.CriticalChance, StatType.CriticalDamage
    ]

    const count = rarity - 1

    for (let i = 0; i < count; i++) {
        const index = rangeInt(0, pool.length)

        const type: StatType = pool[index]!
        pool.splice(index, 1)

        const value = getSubStatRange(type).roll()

        subStats.push({ statType: type, statValue: value })
    }

    return subStats
}

function getRandomEquipmentType(): EquipmentType {
    const values = Object.values(EquipmentType)
    const index = rangeInt(0, values.length)

    const type = values[index] as EquipmentType

    return type
}

function generateRating(mainStatValue: number, subStats: SubStat[], rarity: number): number {
  const scores: number[] = [];

  const mainStatRange = getMainStatRange(rarity);
  scores.push(mainStatRange.normalize(mainStatValue) * 100);

  for (const sub of subStats) {
    const subStatRange = getSubStatRange(sub.statType);
    scores.push(subStatRange.normalize(sub.statValue) * 100);
  }

  const average = scores.reduce((sum, s) => sum + s, 0) / scores.length;
  return Math.round(average);
}